File: CrackingTheCodingInterview/Ch1/ch1_4.py
import unittest


def IsPerm(char):

    if len(char) % 2 == 0:
        count = len(char) // 2
    else:
        count = (len(char) - 1) // 2

    for i in range(count):
        pointer_head = char[i]
        pointer_tail = char[-(i + 1)]
        if pointer_head != pointer_tail:
            return False
    return True



class Test(unittest.TestCase):

    def test_IsPerm(self):
        dataT = ['akasaka', 'abccba']
        dataF = ['roppongi', 'abcabc']

        for t in dataT:
            actual = IsPerm(t)
            self.assertTrue(actual)

        for t in dataF:
            actual = IsPerm(t)
            self.assertFalse(actual)

if __name__ == '__main__':
    unittest.main(argv=['first-arg-is-ignored'], exit=False)


#javaと同じ書き方！
def Isperm(phrase):
    table = [ord(i) for i in phrase]
    return foundodd(table)

def foundodd(table):
    odd = False
    for i in table:
        if i % 2 == 0:
            if odd:
                return False
            odd = True
        return True

File: CrackingTheCodingInterview/Ch1/ch1_4.java
class Main{
  public static void main(String[] args){
    String str1 = "akasaka";
    String str2 = "minato";
    output(str1);
    output(str2);
  }

  public static void output(String str){
    System.out.println(str);
    System.out.println(IsPerm(str));
  }

  public static boolean IsPerm(String phrase){
    int[] table = buildCharFreq(phrase);
    return checkMaxOneOdd(table);
  }

  public static boolean checkMaxOneOdd(int[] table){
    boolean foundOdd = false;
    for (int count: table){
      if (count % 2 == 1){
        if (foundOdd){
          return false;
        }
        foundOdd = true;
      }
    }
    return true;
  }

  public static int getCharNum(Character c){
    int a = Character.getNumericValue('a');
    int z = Character.getNumericValue('z');
    int val = Character.getNumericValue(c);
    if (a <= val && val <= z){
      return val - a;
    }
    return -1;
  }

  public static int[] buildCharFreq(String phrase){
    int[] table = new int[Character.getNumericValue('z') - Character.getNumericValue('a') + 1];
    for (char c : phrase.toCharArray()){
      int x = getCharNum(c);
      if (x != 1){
        table[x] ++;
      }
    }
    return table;
  }

}
